### PR TITLE
linkerd integration with cert-manager

### PR DESCRIPTION
**Pull request by mandsa01**

### CODE REVIEW

File: charts/linkerd-issuers/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+apiVersion: v2
```
Comment (ggolub) [inline charts/linkerd-issuers/Chart.yaml:1]: Why are we adding a helm chart for the certificates when we already have pure terraform code that creates them?  The terraform code would need enhancements for the additional linkerd helm charts (linkerd-viz, linkerd-jaeger) but I don't see why we abandon the original approach.

File: main.tf [hunk 2]
```diff
@@ -24,58 +26,150 @@ resource "tls_self_signed_cert" "linkerd-ca" {
...
-resource "helm_release" "linkerd" {
-  name             = "linkerd"
-  chart            = "${var.chart_repository}/linkerd2-${var.chart_version}.tgz"
-  version          = var.chart_version
-  create_namespace = true
 
-  set {
-    name  = "global.identityTrustAnchorsPEM"
-    value = tls_self_signed_cert.linkerd-ca.cert_pem
+resource "kubernetes_secret" "webhook_trust_anchor" {
```
Comment (ggolub) [inline main.tf:54]: Why are we creating secrets in K8s with this information?  Is that required by the helm chart?

**Review by ggolub — CHANGES_REQUESTED** [review]
Timing of this is good as I was going to update the module to use the newer linkerd version.
However, there are massive changes and I don't understand the reasons behind all of them.

Major concerns:
- Why are we adding a helm chart to manage the certificates when we already had existing "pure" terrform code that did that?
- Lots of renaming/reorganization that adds confusion to the review and anyone using the existing module but I don't see the benefit of it.

Additional, inline, comments for specific things in the code.

In general, I think the original module should be enhanced to account for linkerd's splitting of their helm chart into multiple charts.  I don't think we should be bundling helm certificate management into this when the terraform code works.

**Comment by mandsa01** [comment]
> Timing of this is good as I was going to update the module to use the newer linkerd version.
> However, there are massive changes and I don't understand the reasons behind all of them.
> 
> Major concerns:
> 
> * Why are we adding a helm chart to manage the certificates when we already had existing "pure" terrform code that did that?
> * Lots of renaming/reorganization that adds confusion to the review and anyone using the existing module but I don't see the benefit of it.
> 
> Additional, inline, comments for specific things in the code.
> 
> In general, I think the original module should be enhanced to account for linkerd's splitting of their helm chart into multiple charts. I don't think we should be bundling helm certificate management into this when the terraform code works.



> Timing of this is good as I was going to update the module to use the newer linkerd version.
> However, there are massive changes and I don't understand the reasons behind all of them.
> 
> Major concerns:
> 
> * Why are we adding a helm chart to manage the certificates when we already had existing "pure" terrform code that did that?
> * Lots of renaming/reorganization that adds confusion to the review and anyone using the existing module but I don't see the benefit of it.
> 
> Additional, inline, comments for specific things in the code.
> 
> In general, I think the original module should be enhanced to account for linkerd's splitting of their helm chart into multiple charts. I don't think we should be bundling helm certificate management into this when the terraform code works.

Linkerd automatically rotates the TLS certificates for data plane proxies every 24 hours, it does not rotate the TLS credentials used to issue these certificate, issuer certificate and private key are generated by using cert-manager

https://linkerd.io/2.10/tasks/automatically-rotating-control-plane-tls-credentials/

https://linkerd.io/2.10/tasks/automatically-rotating-webhook-tls-credentials/

File: main.tf [hunk 3]
```diff
@@ -1,17 +1,20 @@
 locals {
-  # set certification expiration date for the number of hours specified
   cert_expiration_date = timeadd(timestamp(), "${var.ca_cert_expiration_hours}h")
+  namespace = toset([
+    "linkerd",
+    "linkerd-viz",
+    "linkerd-jaeger"
+  ])
```
[the remark this answers is not in the document]
Comment (mandsa01) [reply main.tf:7]: i'll use a variable to install Jaeger

File: variables.tf [hunk 4]
```diff
@@ -19,16 +19,34 @@ variable "trust_anchor_validity_hours" {
...
+  description = "Number of hours added to installation time to calculate trust anchor certification expiration date"
+  type        = string
+  default     = "24h"
+}
+
+variable "certificate_webhook_renewbefore" {
+  description = "Number of hours added to installation time to calculate trust anchor certification expiration date"
+  type        = string
+  default     = "1h"
 }
```
[the remark this answers is not in the document]
Comment (mandsa01) [reply variables.tf:50]: i'll rework on the description.

File: main.tf [hunk 5]
```diff
@@ -24,58 +27,151 @@ resource "tls_self_signed_cert" "linkerd-ca" {
...
+    var.additional_yaml_config
+  ]
+}
+
+resource "helm_release" "linkerd-jaeger" {
+  depends_on = [helm_release.linkerd-viz]
+  name       = "linkerd-jaeger"
+  repository = "https://helm.linkerd.io/edge"
+  chart      = "linkerd-jaeger"
+  version    = "21.3.4"
```
[the remark this answers is not in the document]
Comment (mandsa01) [reply main.tf:165]: i'll remove the hard-code

File: main.tf [hunk 6]
```diff
@@ -24,58 +27,151 @@ resource "tls_self_signed_cert" "linkerd-ca" {
...
-  validity_period_hours = var.issuer_validity_hours
-  allowed_uses          = ["cert_signing", "crl_signing"]
+resource "kubernetes_namespace" "namespace" {
+  for_each = local.namespace
+  metadata {
+    name = each.key
+    annotations = {
+      "linkerd.io/inject" = "enabled"
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mandsa01) [reply main.tf:51]: viz and Jaeger need those side car proxies

File: main.tf (hunk 2 above)
Comment (mandsa01) [reply main.tf:54]: need to create the secrets before installing the helm charts

File: main.tf (hunk 6 above)
[the remark this answers is not in the document]
Comment (ggolub) [reply main.tf:51]: I don't see the annotation in linkerd-viz (haven't deployed linkerd-jaeger).  Can you show me where that comes from?  I thought it was used to inject into non-linkerd namespaces.

**Comment by ggolub** [comment]
Can terraform code be used to set up the rotating credentials instead of using a helm chart?

**Pull request closed without merging**